File: _examples/helloworld/native/android/app/src/main/java/com/wailsplugin/WailsPlugin.java
package com.wailsplugin;

import android.content.Context;
import android.content.Intent;

/**
 * WailsPlugin is the base interface for all native Android extensions.
 */
public interface WailsPlugin {
    
    String getDomain();
    
    /**
     * Called when the plugin is registered or when an activity becomes active.
     * @param context Can be an Application context or an AppCompatActivity context.
     */
    void onAttach(Context context);
    
    String handleAction(String action, String jsonArgsPayload);
    
    void onActivityResult(int requestCode, int resultCode, Intent data);
    
    void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults);
}

File: _examples/helloworld/native/ios/WailsPlugin.swift
import UIKit

/// WailsPlugin is the iOS equivalent of the Android `com.wailsplugin.WailsPlugin` interface.
public protocol WailsPlugin {
    /// Unique plugin domain used by Go-to-native calls, e.g. "permissions".
    var domain: String { get }

    /// Called when the plugin is attached to the top-level view controller.
    func onAttach(_ viewController: UIViewController)

    /// Called by the native handler when Go needs the plugin to perform an action.
    func handleAction(_ action: String, jsonArgsPayload: String) -> String

    /// Optional lifecycle hook for deep links or open URL events.
    func onOpenURL(_ url: URL)

    /// Optional lifecycle hook for permission results.
    func onPermissionResult(_ requestCode: Int, permissions: [String], grantResults: [Int])
}

public extension WailsPlugin {
    func onAttach(_ viewController: UIViewController) {}
    func onOpenURL(_ url: URL) {}
    func onPermissionResult(_ requestCode: Int, permissions: [String], grantResults: [Int]) {}
}
